extension/src/bs/BspProxy.ts

```diff
@@ -0,0 +1,95 @@
+
+    constructor(context: vscode.ExtensionContext, private readonly logger: Logger) {
+        this.buildServerConnector = new BuildServerConnector();
+        this.jdtlsImporterConnector = new JdtlsImporterConnector(context);
+    }
+    /**
+     * This function needs to be called before we start Java Gradle Server.
+     */
+    public prepareToStart(): void {
+        this.buildServerConnector.setupBuildServerPipeStream();
+    }
+
+    /**
+     * The order of the following start steps is important.
+     *
+     * We have to start listening after the message forwarding is setup, otherwise the Java importer
+     * will stop polling and start sending messages before the forwarding is setup and the messages will be lost.
+     */
+    public async start(): Promise<void> {
+        await this.jdtlsImporterConnector.waitForImporterPipePath();
+        await this.jdtlsImporterConnector.setupImporterPipeStream();
+
+        this.setupMessageForwarding(
+            this.jdtlsImporterConnector.getImporterConnection(),
+            this.buildServerConnector.getServerConnection()
+        );
+        this.jdtlsImporterConnector.startListening();
+    }
+
+    public getBuildServerPipeName(): string {
+        return this.buildServerConnector.getServerPipePath();
+    }
+
+    private setupMessageForwarding(
+        importerConnection: rpc.MessageConnection | null,
+        buildServerConnection: rpc.MessageConnection | null
+    ): void {
+        importerConnection?.onRequest((method, params) => {
+            if (params !== null) {
+                return buildServerConnection?.sendRequest(method, params);
+            }
+            return buildServerConnection?.sendRequest(method);
+        });
+
+        buildServerConnection?.onNotification((method, params) => {
+            if (params !== null) {
+                return importerConnection?.sendNotification(method, params);
+            }
+            importerConnection?.sendNotification(method);
+        });
+        importerConnection?.onError(([error, message, code]) => {
+            this.logger.error(`Error on importerConnection: ${error.message}`);
+            sendInfo("", {
+                kind: "bspProxy-importerConnectionError",
+                data2: JSON.stringify({
+                    error,
+                    message,
+                    code,
+                }),
+            });
+            // TODO: Implement more specific error handling logic here
+        });
+
+        buildServerConnection?.onError(([error, message, code]) => {
+            this.logger.error(`Error on buildServerConnection: ${error.message}`);
+            sendInfo("", {
+                kind: "bspProxy-importerConnectionError",
+                data2: JSON.stringify({
+                    error,
+                    message,
+                    code,
+                }),
+            });
+            // TODO: Implement more specific error handling logic here
+        });
```

Comment:
`rpc.MessageConnection` provide an `onError` listerner. I'm not sure when & how this connection might occur error, so I think we can collect Telemetry here to get more insight